Detectors/TPC/calibration/SpacePoints/src/ResidualAggregator.cxx

```diff
@@ -0,0 +1,205 @@
+          auto& statsVoxel = stats[iSec][trackResiduals->getGlbVoxBin(ix, ip, iz)];
+          // COG estimates are set to the bin center by default
+          trackResiduals->getVoxelCoordinates(iSec, ix, ip, iz, statsVoxel.meanPos[TrackResiduals::VoxX], statsVoxel.meanPos[TrackResiduals::VoxF], statsVoxel.meanPos[TrackResiduals::VoxZ]);
+        }
+      }
+    }
+    treeOutResiduals->Branch(Form("sec%d", iSec), &residualsPtr[iSec]);
+    treeOutStats->Branch(Form("sec%d", iSec), &statsPtr[iSec]);
+  }
+}
+
+void ResidualsContainer::fillStatisticsBranches()
+{
+  // only called when the slot is finalized, otherwise treeOutStats
+  // remains empty and we keep the statistics in memory in the vectors
+  // (since their size anyway does not change)
+  treeOutStats->Fill();
+}
+
+void ResidualsContainer::fill(const gsl::span<const TrackResiduals::UnbinnedResid> data)
+{
+  // receives large vector of unbinned residuals and fills the sector-wise vectors
+  // with binned residuals and statistics
+  LOG(debug) << "Filling ResidualsContainer with vector of size " << data.size();
+  for (const auto& residIn : data) {
+    int sec = residIn.sec;
+    auto& residVecOut = residuals[sec];
+    auto& statVecOut = stats[sec];
+    std::array<unsigned char, TrackResiduals::VoxDim> bvox;
+    float yPos = residIn.y * param::MaxY / 0x7fff;
+    float zPos = residIn.z * param::MaxZ / 0x7fff;
+    if (!trackResiduals->findVoxelBin(sec, param::RowX[residIn.row], yPos, zPos, bvox)) {
+      // we are not inside any voxel
+      LOGF(debug, "Dropping residual in sec(%i), x(%f), y(%f), z(%f)", sec, param::RowX[residIn.row], yPos, zPos);
+      continue;
+    }
+    residVecOut.emplace_back(residIn.dy, residIn.dz, residIn.tgSlp, bvox);
+    auto& stat = statVecOut[trackResiduals->getGlbVoxBin(bvox)];
+    float& binEntries = stat.nEntries;
+    float oldEntries = binEntries++;
+    float norm = 1.f / binEntries;
+    // update COG for voxel bvox (don't need to update X here, it stays at the pad row radius)
+    stat.meanPos[TrackResiduals::VoxF] = (stat.meanPos[TrackResiduals::VoxF] * oldEntries + yPos / param::RowX[residIn.row]) * norm;
+    stat.meanPos[TrackResiduals::VoxZ] = (stat.meanPos[TrackResiduals::VoxZ] * oldEntries + zPos / param::RowX[residIn.row]) * norm;
+    ++nResidualsTotal;
+  }
+  treeOutResiduals->Fill();
+  for (auto& residVecOut : residuals) {
+    residVecOut.clear();
+  }
+}
+
+void ResidualsContainer::merge(ResidualsContainer* prev)
+{
+  // the previous slot is merged to this one and afterwards
+  // the previous one will be deleted
+  for (int iSec = 0; iSec < SECTORSPERSIDE * SIDES; ++iSec) {
+    // merge statistics
+    const auto& statVecPrev = prev->stats[iSec];
+    auto& statVec = stats[iSec];
+    for (int iVox = 0; iVox < trackResiduals->getNVoxelsPerSector(); ++iVox) {
+      const auto& statPrev = statVecPrev[iVox];
+      auto& stat = statVec[iVox];
+      float norm = 1. / (statPrev.nEntries + stat.nEntries);
```

Comment:
one should probably have a protection against division by 0 here.

Reply:
Done